samcli/lib/bootstrap/companion_stack/companion_stack_manager.py

```diff
@@ -0,0 +1,262 @@
+        self._s3_prefix = s3_prefix
+        try:
+            self._cfn_client = boto3.client("cloudformation", config=self._boto_config)
+            self._ecr_client = boto3.client("ecr", config=self._boto_config)
+            self._s3_client = boto3.client("s3", config=self._boto_config)
+            self._account_id = boto3.client("sts").get_caller_identity().get("Account")
+            self._region_name = self._cfn_client.meta.region_name
+        except NoCredentialsError as ex:
+            raise CredentialsError(
+                "Error Setting Up Managed Stack Client: Unable to resolve "
+                "credentials for the AWS SDK for Python client. "
+                "Please see their documentation for options to pass in credentials: "
+                "https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html"
+            ) from ex
+        except NoRegionError as ex:
+            raise RegionError(
+                "Error Setting Up Managed Stack Client: Unable to resolve a region. "
+                "Please provide a region via the --region parameter or by the AWS_REGION environment variable."
+            ) from ex
+
+    def set_functions(self, function_logical_ids: List[str]) -> None:
+        """
+        Sets functions that need to have ECR repos created
+
+        Parameters
+        ----------
+        function_logical_ids: List[str]
+            Function logical IDs that need to have ECR repos created
+        """
+        self._builder.clear_functions()
+        for function_logical_id in function_logical_ids:
+            self._builder.add_function(function_logical_id)
+
+    def update_companion_stack(self) -> None:
+        """
+        Blocking call to create or update the companion stack based on current functions
+        Companion stack template will be updated to the s3 bucket first before deployment
+        """
+        stack_name = self._companion_stack.stack_name
+        template = self._builder.build()
+
+        with mktempfile() as temporary_file:
+            temporary_file.write(template)
+            temporary_file.flush()
+
+            s3_uploader = S3Uploader(
+                self._s3_client, bucket_name=self._s3_bucket, prefix=self._s3_prefix, no_progressbar=True
+            )
+            # TemplateUrl property requires S3 URL to be in path-style format
+            parts = S3Uploader.parse_s3_url(
+                s3_uploader.upload_with_dedup(temporary_file.name, "template"), version_property="Version"
+            )
+
+        template_url = s3_uploader.to_path_style_s3_url(parts["Key"], parts.get("Version", None))
+
+        exists = self.does_companion_stack_exist()
+        if exists:
+            self._cfn_client.update_stack(
+                StackName=stack_name, TemplateURL=template_url, Capabilities=["CAPABILITY_AUTO_EXPAND"]
+            )
+            waiter = self._cfn_client.get_waiter("stack_update_complete")
+        elif self._builder.repo_mapping:
```

Comment:
is this `if` condition necessary? We have no `else` here, waiter at line 116 could be `None`.

Reply:
Removed `elif` and added `repo_mapping` check at the beginning.